In [2]:
#Load Libraries
import pandas as pd
import numpy as np

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:


In [84]:
#Read Postal Codes of Canada into dataframe

df=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M') 
df =df[0]
df.rename(columns={'Postal Code':'PostalCode'}, inplace=True)
df.head()
 

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

In [85]:
df.set_index("PostalCode" , inplace=True)
df.head()
 

Borough               Neighborhood
PostalCode                                             
M1A             Not assigned               Not assigned
M2A             Not assigned               Not assigned
M3A               North York                  Parkwoods
M4A               North York           Victoria Village
M5A         Downtown Toronto  Regent Park, Harbourfront

More than one neighborhood can exist in one postal code area.

In [86]:
#Drop Neighborhood is no assigned

df=df[df["Neighborhood"]!='Not assigned']
df.head()

Borough                                 Neighborhood
PostalCode                                                               
M3A               North York                                    Parkwoods
M4A               North York                             Victoria Village
M5A         Downtown Toronto                    Regent Park, Harbourfront
M6A               North York             Lawrence Manor, Lawrence Heights
M7A         Downtown Toronto  Queen's Park, Ontario Provincial Government

if a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [87]:
#reemplazar values

import numpy as np
df['Borough'] = np.where(df['Borough'] =="Not assigned", df['Neighborhood'], df['Borough'])
df.head()

Borough                                 Neighborhood
PostalCode                                                               
M3A               North York                                    Parkwoods
M4A               North York                             Victoria Village
M5A         Downtown Toronto                    Regent Park, Harbourfront
M6A               North York             Lawrence Manor, Lawrence Heights
M7A         Downtown Toronto  Queen's Park, Ontario Provincial Government

In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [88]:
df.shape

(103, 2)

In [164]:
df_geo=pd.read_csv("http://cocl.us/Geospatial_data")
df_geo.rename(columns={'Postal Code':'PostalCode'}, inplace=True)
df_geo.set_index("PostalCode" , inplace=True)
df_geo.head()
 


Latitude  Longitude
PostalCode                      
M1B         43.806686 -79.194353
M1C         43.784535 -79.160497
M1E         43.763573 -79.188711
M1G         43.770992 -79.216917
M1H         43.773136 -79.239476

In [168]:
neighborhoods= pd.merge(df,df_geo, on='PostalCode', how='inner')
neighborhoods.head()

Borough                                 Neighborhood  \
PostalCode                                                                  
M3A               North York                                    Parkwoods   
M4A               North York                             Victoria Village   
M5A         Downtown Toronto                    Regent Park, Harbourfront   
M6A               North York             Lawrence Manor, Lawrence Heights   
M7A         Downtown Toronto  Queen's Park, Ontario Provincial Government   

             Latitude  Longitude  
PostalCode                        
M3A         43.753259 -79.329656  
M4A         43.725882 -79.315572  
M5A         43.654260 -79.360636  
M6A         43.718518 -79.464763  
M7A         43.662301 -79.389494

In [169]:
neighborhoods.shape

(103, 4)

make sure that the dataset has all 10 boroughs and 103 neighborhoods.

In [170]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [171]:
#### Use geopy library to get the latitude and longitude values of New York City.

In [173]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [174]:
#### Create a map of Toronto with neighborhoods superimposed on top.

In [175]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Downtown Toronto data.

In [183]:
toronto_data = neighborhoods[neighborhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
toronto_data.head()

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  
0 -79.360636  
1 -79.389494  
2 -79.378937  
3 -79.375418  
4 -79.373306

let's get the geographical coordinates of Downtown Toronto.

In [178]:
address = 'Downtown Toronto, Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


As we did with all of New York City, let's visualizat Downtown Toronto the neighborhoods in it.

In [184]:
# create map of Manhattan using latitude and longitude values
map_dt = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_dt)  
    
map_dt

#### Define Foursquare Credentials and Version

In [186]:
CLIENT_ID = 'YAISAMGMDRYD5TB0EQAE0XKDPFFI1LQK3NO4KBRQKS3E1S3I' # your Foursquare ID
CLIENT_SECRET = '4UHOSKSK5CJZTL3EMORA200ZVNZLI4AQAM5BY1WIYLHDSCBO' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
 

Your credentails:
CLIENT_ID: YAISAMGMDRYD5TB0EQAE0XKDPFFI1LQK3NO4KBRQKS3E1S3I
CLIENT_SECRET:4UHOSKSK5CJZTL3EMORA200ZVNZLI4AQAM5BY1WIYLHDSCBO


Get the neighborhood's name.

In [187]:
toronto_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

Get the neighborhood's latitude and longitude values.

In [188]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


In [189]:
#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

In [191]:
# type your answer here
radius=500
LIMIT = 10 # limit of number of venues returned by Foursquare API
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
  # display URL
url

'https://api.foursquare.com/v2/venues/explore?&client_id=YAISAMGMDRYD5TB0EQAE0XKDPFFI1LQK3NO4KBRQKS3E1S3I&client_secret=4UHOSKSK5CJZTL3EMORA200ZVNZLI4AQAM5BY1WIYLHDSCBO&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=10'

Send the GET request and examine the resutls

In [192]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f10f7eb72fc2b16d130e964'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 46,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [194]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [196]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4          Impact Kitchen           Restaurant  43.656369 -79.356980

In [201]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

10 venues were returned by Foursquare.


## Explore Neighborhoods in Manhattan

In [200]:
#### Let's create a function to repeat the same process to all the neighborhoods in toronto

In [203]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [205]:
#### Now write the code to run the above function on each neighborhood and create a new dataframe called *manhattan_venues*.

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )



Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


Let's check how many venues were returned for each neighborhood

In [206]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            10   
CN Tower, King and Spadina, Railway Lands, Harb...                     10   
Central Bay Street                                                     10   
Christie                                                               10   
Church and Wellesley                                                   10   
Commerce Court, Victoria Hotel                                         10   
First Canadian Place, Underground city                                 10   
Garden District, Ryerson                                               10   
Harbourfront East, Union Station, Toronto Islands                      10   
Kensington Market, Chinatown, Grange Park                              10   
Queen's Park, Ontario Provincial Government                            10   
Regent Park, Harbourfront                                              10   
Richmond, Adelaide, King                                               10   
Rosedale                                                                4   
St. James Town                                                         10   
St. James Town, Cabbagetown                                            10   
Stn A PO Boxes                                                         10   
Toronto Dominion Centre, Design Exchange                               10   
University of Toronto, Harbord                                         10   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             10   
CN Tower, King and Spadina, Railway Lands, Harb...                      10   
Central Bay Street                                                      10   
Christie                                                                10   
Church and Wellesley                                                    10   
Commerce Court, Victoria Hotel                                          10   
First Canadian Place, Underground city                                  10   
Garden District, Ryerson                                                10   
Harbourfront East, Union Station, Toronto Islands                       10   
Kensington Market, Chinatown, Grange Park                               10   
Queen's Park, Ontario Provincial Government                             10   
Regent Park, Harbourfront                                               10   
Richmond, Adelaide, King                                                10   
Rosedale                                                                 4   
St. James Town                                                          10   
St. James Town, Cabbagetown                                             10   
Stn A PO Boxes                                                          10   
Toronto Dominion Centre, Design Exchange                                10   
University of Toronto, Harbord                                          10   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            10              10   
CN Tower, King and Spadina, Railway Lands, Harb...     10              10   
Central Bay Street                                     10              10   
Christie                                               10              10   
Church and Wellesley                                   10              10   
Commerce Court, Victoria Hotel                         10              10   
First Canadian Place, Underground city                 10              10   
Garden District, Ryerson                               10              10   
Harbourfront East, Union 

## Analyze Each Neighborhood

In [207]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  \
0                0                 0                    0   
1                0                 0                    0   
2                0                 0                    0   
3                0                 0                    0   
4                0                 0                    0   

   Arts & Crafts Store  Bakery  Bar  Beer Bar  Bookstore  Boutique  \
0                    0       1    0         0          0         0   
1                    0       0    0         0          0         0   
2                    0       0    0         0          0         0   
3                    0       0    0         0          0         0   
4                    0       0    0         0          0         0   

   Breakfast Spot  Bubble Tea Shop  Burrito Place  Café  Candy Store  \
0               0                0              0     0            0   
1               0                0              0     0            0   
2               0                0              0     0            0   
3               0                0              0     0            0   
4               0                0              0     0            0   

   Caribbean Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  \
0                     0               0             0            0   
1                     0               0             0            1   
2                     0               0             0            0   
3                     0               0             0            0   
4                     0               0             0            0   

   Comic Shop  Concert Hall  Cosmetics Shop  Creperie  Dance Studio  \
0           0             0               0         0             0   
1           0             0               0         0             0   
2           0             0               0         0             0   
3           0             0               0         0             0   
4           0             0               0         0             0   

   Dessert Shop  Diner  Distribution Center  Farmers Market  Food Truck  \
0             0      0                    0               0           0   
1             0      0                    0               0           0   
2             0      0                    1               0           0   
3             0      0                    0               0           0   
4             0      0                    0               0           0   

   Fountain  Gastropub  General Entertainment  Grocery Store  Gym  \
0         0          0                      0              0    0   
1         0          0                      0              0    0   
2         0          0                      0              0    0   
3         0          0                      0              0    0   
4         0          0                      0              0    0   

   Gym / Fitness Center  Harbor / Marina  Historic Site  Hotel  \
0                     0                0              0      0   
1                     0                0              0      0   
2                     0                0              0      0   
3                     0                0              0      0   
4                     0                0              0      0   

   Indian Restaurant  Italian Restaurant  Japanese Restaurant  Jewelry Store  \
0                  0                   0                    0              0   
1                  0                   0        

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [210]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  Airport  \
0                                         Berczy Park          0.0      0.0   
1   CN Tower, King and Spadina, Railway Lands, Har...          0.0      0.1   
2                                  Central Bay Street          0.0      0.0   
3                                            Christie          0.0      0.0   
4                                Church and Wellesley          0.0      0.0   
5                      Commerce Court, Victoria Hotel          0.0      0.0   
6              First Canadian Place, Underground city          0.0      0.0   
7                            Garden District, Ryerson          0.0      0.0   
8   Harbourfront East, Union Station, Toronto Islands          0.0      0.0   
9           Kensington Market, Chinatown, Grange Park          0.0      0.0   
10        Queen's Park, Ontario Provincial Government          0.1      0.0   
11                          Regent Park, Harbourfront          0.0      0.0   
12                           Richmond, Adelaide, King          0.0      0.0   
13                                           Rosedale          0.0      0.0   
14                                     St. James Town          0.0      0.0   
15                        St. James Town, Cabbagetown          0.0      0.0   
16                                     Stn A PO Boxes          0.0      0.0   
17           Toronto Dominion Centre, Design Exchange          0.0      0.0   
18                     University of Toronto, Harbord          0.1      0.0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                  0.0           0.0             0.0              0.0   
1                  0.1           0.1             0.2              0.1   
2                  0.0           0.0             0.0              0.0   
3                  0.0           0.0             0.0              0.0   
4                  0.0           0.0             0.0              0.0   
5                  0.0           0.0             0.0              0.0   
6                  0.0           0.0             0.0              0.0   
7                  0.0           0.0             0.0              0.0   
8                  0.0           0.0             0.0              0.0   
9                  0.0           0.0             0.0              0.0   
10                 0.0           0.0             0.0              0.0   
11                 0.0           0.0             0.0              0.0   
12                 0.0           0.0             0.0              0.0   
13                 0.0           0.0             0.0              0.0   
14                 0.0           0.0             0.0              0.0   
15                 0.0           0.0             0.0              0.0   
16                 0.0           0.0             0.0              0.0   
17                 0.0           0.0             0.0              0.0   
18                 0.0           0.0             0.0              0.0   

    Airport Terminal  American Restaurant  Arts & Crafts Store  Bakery  Bar  \
0                0.0                  0.0                  0.0     0.0  0.0   
1                0.1                  0.0                  0.0     0.0  0.0   
2                0.0                  0.0                  0.0     0.0  0.0   
3                0.0                  0.0                  0.0     0.0  0.0   
4                0.0                  0.0                  0.0     0.0  0.0   
5                0.0                  0.0                  0.0     0.1  0.0   
6                0.0                  0.1                  0.0     0.1  0.0   
7                0.0                  0.0                  0.0     0.0  0.0   
8                0.0                  0.0                  0.0     0.0  0.0   
9                0.0                  0.0                  0.1     0.1  0.0   
10               0.0                  0.0                  0.1     0.0  0.0   
11               0.0                  0.0          

#### Let's print each neighborhood along with the top 5 most common venues

In [211]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0      Beer Bar   0.1
1  Liquor Store   0.1
2          Park   0.1
3  Concert Hall   0.1
4        Museum   0.1


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0      Airport Lounge   0.2
1  Airport Food Court   0.1
2        Airport Gate   0.1
3     Airport Service   0.1
4    Airport Terminal   0.1


----Central Bay Street----
                 venue  freq
0          Coffee Shop   0.3
1          Pizza Place   0.1
2                  Spa   0.1
3            Gastropub   0.1
4  Japanese Restaurant   0.1


----Christie----
                venue  freq
0                Café   0.3
1       Grocery Store   0.2
2  Italian Restaurant   0.1
3         Candy Store   0.1
4          Restaurant   0.1


----Church and Wellesley----
                venue  freq
0            Beer Bar   0.1
1  Mexican Restaurant   0.1
2        Dance Studio   0.1
3    Ramen Restaurant   0.1
4 

In [212]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [215]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park                  Park   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie                  Café   
4                               Church and Wellesley                  Park   

  2nd Most Common Venue 3rd Most Common Venue       4th Most Common Venue  \
0          Concert Hall          Cocktail Bar                Liquor Store   
1               Airport    Airport Food Court                Airport Gate   
2             Gastropub      Sushi Restaurant  Modern European Restaurant   
3         Grocery Store           Candy Store                  Restaurant   
4            Restaurant    Mexican Restaurant            Theme Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Beer Bar            Restaurant                Museum   
1            Airport Service      Airport Terminal       Harbor / Marina   
2  Middle Eastern Restaurant           Pizza Place   Japanese Restaurant   
3                Coffee Shop                 Diner    Italian Restaurant   
4            Bubble Tea Shop        Breakfast Spot          Dance Studio   

  8th Most Common Venue          9th Most Common Venue 10th Most Common Venue  
0       Thai Restaurant  Vegetarian / Vegan Restaurant         Farmers Market  
1           Coffee Shop                       Boutique  Vietnamese Restaurant  
2                   Spa                   Dance Studio               Creperie  
3          Cocktail Bar                     Comic Shop           Concert Hall  
4             Bookstore                       Beer Bar       Ramen Restaurant

## Cluster Neighborhoods

In [217]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 2, 2, 1, 0, 3, 3, 1, 0, 1])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [218]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               3                  Park                Bakery   
1 -79.389494               2           Coffee Shop           Yoga Studio   
2 -79.378937               1           Music Venue                  Café   
3 -79.375418               2           Coffee Shop                   Gym   
4 -79.373306               0                  Park          Concert Hall   

  3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0        Breakfast Spot                Coffee Shop                   Pub   
1      Sushi Restaurant                   Creperie    Italian Restaurant   
2               Theater            Thai Restaurant              Tea Room   
3    Italian Restaurant  Middle Eastern Restaurant            Restaurant   
4          Cocktail Bar               Liquor Store              Beer Bar   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Restaurant                   Spa         Historic Site   
1   Arts & Crafts Store   Distribution Center                  Park   
2           Pizza Place        Clothing Store                 Plaza   
3        Cosmetics Shop   Japanese Restaurant              Creperie   
4            Restaurant                Museum       Thai Restaurant   

           9th Most Common Venue 10th Most Common Venue  
0           Gym / Fitness Center    Distribution Center  
1                          Diner           Dessert Shop  
2                  Burrito Place             Comic Shop  
3                     Food Truck           Dance Studio  
4  Vegetarian / Vegan Restaurant         Farmers Market

In [219]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

# Cluster 1

In [224]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
4                                         Berczy Park                  Park   
8   Harbourfront East, Union Station, Toronto Islands                 Plaza   
11                     University of Toronto, Harbord           Yoga Studio   
15                                     Stn A PO Boxes                  Park   
18                               Church and Wellesley                  Park   

   2nd Most Common Venue  3rd Most Common Venue  \
4           Concert Hall           Cocktail Bar   
8                   Park  Performing Arts Venue   
11    Italian Restaurant              Bookstore   
15            Restaurant             Food Truck   
18            Restaurant     Mexican Restaurant   

            4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
4                    Liquor Store              Beer Bar            Restaurant   
8                    Dessert Shop                 Hotel   Sporting Goods Shop   
11                       Beer Bar            Restaurant                   Bar   
15  Vegetarian / Vegan Restaurant          Concert Hall          Cocktail Bar   
18               Theme Restaurant       Bubble Tea Shop        Breakfast Spot   

   7th Most Common Venue 8th Most Common Venue          9th Most Common Venue  \
4                 Museum       Thai Restaurant  Vegetarian / Vegan Restaurant   
8                   Lake           Salad Place                     Comic Shop   
11   Japanese Restaurant                Bakery               Sushi Restaurant   
15                Museum              Beer Bar                       Fountain   
18          Dance Studio             Bookstore                       Beer Bar   

   10th Most Common Venue  
4          Farmers Market  
8    Caribbean Restaurant  
11           Dessert Shop  
15        Thai Restaurant  
18       Ramen Restaurant

# Cluster 2

In [227]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
2                    Garden District, Ryerson           Music Venue   
6                                    Christie                  Café   
12  Kensington Market, Chinatown, Grange Park                  Café   
16                St. James Town, Cabbagetown                  Café   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                    Café               Theater       Thai Restaurant   
6           Grocery Store           Candy Store            Restaurant   
12  Vietnamese Restaurant   Arts & Crafts Store          Cocktail Bar   
16  General Entertainment     Indian Restaurant            Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2               Tea Room           Pizza Place        Clothing Store   
6            Coffee Shop                 Diner    Italian Restaurant   
12           Coffee Shop    Mexican Restaurant       Organic Grocery   
16         Jewelry Store   Japanese Restaurant    Italian Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2                  Plaza         Burrito Place             Comic Shop  
6           Cocktail Bar            Comic Shop           Concert Hall  
12                Bakery  Caribbean Restaurant           Dance Studio  
16                Bakery                 Diner           Dance Studio